ui/src/app/views/workspace-list.tsx

```diff
@@ -83,66 +84,108 @@ const WorkspaceCardMenu: React.FunctionComponent<{
     </Clickable>
   </PopupTrigger>;
 };
 
-export class WorkspaceCard extends React.Component<
-    {wp: WorkspacePermissions, userEmail: string, reload: Function},
-    {sharing: boolean, confirmDeleting: boolean, workspaceDeletionError: boolean,
-      bugReportOpen: boolean, bugReportError: string}> {
+interface WorkspaceCardState {
+  bugReportError: string;
+  bugReportOpen: boolean;
+  confirmDeleting: boolean;
+  // Whether this card is busy loading data specific to the workspace.
+  loadingData: boolean;
+  sharing: boolean;
+  // The list of user roles associated with this workspace. Lazily populated
+  // only when the workspace share dialog is opened.
+  userRoles?: UserRole[];
+  workspaceDeletionError: boolean;
+}
+
+interface WorkspaceCardProps {
+  userEmail: string;
+  wp: WorkspacePermissions;
+  // A reload function that can be called by this component to reqeust a refresh
+  // of the workspace contained in this card.
+  reload(): Promise<void>;
+}
 
+export class WorkspaceCard extends React.Component<WorkspaceCardProps, WorkspaceCardState> {
   constructor(props) {
     super(props);
     this.state = {
-      bugReportOpen: false,
       bugReportError: '',
-      sharing: false,
+      bugReportOpen: false,
       confirmDeleting: false,
+      loadingData: false,
+      sharing: false,
+      userRoles: null,
       workspaceDeletionError: false
     };
   }
 
   async deleteWorkspace() {
     const {wp} = this.props;
-    workspacesApi().deleteWorkspace(wp.workspace.namespace, wp.workspace.id).then(() => {
-      this.setState({confirmDeleting: false});
-      this.props.reload();
-    }).catch(() => {
-      this.setState({workspaceDeletionError: true});
+    this.setState({
+      confirmDeleting: false,
+      loadingData: true});
+    try {
+      await workspacesApi().deleteWorkspace(wp.workspace.namespace, wp.workspace.id);
+      this.setState({loadingData: false});
+      await this.props.reload();
+    } catch (e) {
+      this.setState({workspaceDeletionError: true, loadingData: false});
+    }
+  }
+
+  // The function called when the "share" action is called on a workspace card
+  // within the workspaces list.
+  async handleShareAction() {
+    this.setState({
+      loadingData: true
     });
+    const userRolesResponse = await workspacesApi().getFirecloudWorkspaceUserRoles(
+      this.props.wp.workspace.namespace,
+      this.props.wp.workspace.id);
+    // Trigger the sharing dialog to be shown.
+    this.setState({
+      loadingData: false,
+      sharing: true,
+      userRoles: userRolesResponse.items});
   }
 
-  shareWorkspace(): void {
+  async handleShareDialogClose() {
     // Share workspace publishes to current workspace,
     // but here we aren't in the context of a workspace
     // so we need to clear it.
     currentWorkspaceStore.next(undefined);
-    this.setState({sharing: false});
-    this.props.reload();
+    this.setState({
+      sharing: false,
+      userRoles: null});
```

Comment:
Opt nit: I don't think it actually matters if you clear userRoles, because we always reload above, right?

Reply:
No it doesn't matter for functionality, but for cleanliness I figured userRoles should only be populated during the time when it's up-to-date and meant to actually be used. It might become stale e.g. if the workspaces list has been updated but we haven't explicitly reloaded the userRoles state variable yet.